provider/openstack: Acceptance Test Failures

Hello,

I noticed that the Acceptance Tests for the OpenStack provider started failing, so I wanted to fix them up.

Some have been easy enough -- just missing parameters that are now being caught or parameters that need to be Computed, but I'm having trouble with some others...

For example, the `TestAccComputeV2Instance_volumeAttach` runs the following:

```
  resource "openstack_blockstorage_volume_v1" "myvol" {
    name = "myvol"
    size = 1
  }

  resource "openstack_compute_instance_v2" "foo" {
    name = "terraform-test"
    volume {
      volume_id = "${openstack_blockstorage_volume_v1.myvol.id}"
    }
  }
```

and results in:

```
--- FAIL: TestAccComputeV2Instance_volumeAttach (63.72s)
        testing.go:137: Step 0 error: After applying this step, the plan was not empty:

                DIFF:

                UPDATE: openstack_compute_instance_v2.foo
                  security_groups.#:           "1" => "0"
                  security_groups.0:           "default" => ""
                  volume.2224808980.device:    "" => "<computed>"
                  volume.2224808980.id:        "" => "<computed>"
                  volume.2224808980.volume_id: "" => "f450c08e-c081-41af-afb7-55cebf93c24e"

                STATE:

                openstack_blockstorage_volume_v1.myvol:
                  ID = f450c08e-c081-41af-afb7-55cebf93c24e
                  attachment.# = 0
                  availability_zone = nova
                  description =
                  name = myvol
                  region = RegionOne
                  size = 1
                  snapshot_id =
                  source_vol_id =
                  volume_type = None
                openstack_compute_instance_v2.foo:
                  ID = 666daa4c-31c1-4e13-bd54-f4ff760680bd
                  access_ip_v4 = 10.1.0.168
                  access_ip_v6 = [dead:beef:4:1000:f816:3eff:fef4:921c]
                  flavor_id = 1
                  flavor_name = m1.tiny
                  image_id = 5f645a5c-9aae-41ba-897a-dbe47ca7de6d
                  image_name = Ubuntu 14.04
                  name = terraform-test
                  network.# = 1
                  network.0.fixed_ip_v4 = 10.1.0.168
                  network.0.fixed_ip_v6 = [dead:beef:4:1000:f816:3eff:fef4:921c]
                  network.0.mac = fa:16:3e:f4:92:1c
                  network.0.name = nova
                  network.0.port =
                  network.0.uuid =
                  region = RegionOne
                  security_groups.# = 1
                  security_groups.0 = default
                  volume.# = 1
                  volume.1850595663.device = /dev/vdc
                  volume.1850595663.id = f450c08e-c081-41af-afb7-55cebf93c24e
                  volume.1850595663.volume_id = f450c08e-c081-41af-afb7-55cebf93c24e

                  Dependencies:
                    openstack_blockstorage_volume_v1.myvol
```

The security group part is easy enough -- I just need to add the default security group to the resource.

The device and ID are indeed computed, however, the `volume_id` is specified.

It's looking like `TypeSet`s that have a mixture of Computed and specified parameters are being flagged for tests?

If I run the same scenarios outside of the testing framework, everything works and the final `terraform show` output is correct.

Is there something I can do to get these types of test to work correctly?

--- a/builtin/providers/openstack/resource_openstack_blockstorage_volume_v1.go
+++ b/builtin/providers/openstack/resource_openstack_blockstorage_volume_v1.go
@@ -53,6 +53,7 @@ func resourceBlockStorageVolumeV1() *schema.Resource {
 				Type:     schema.TypeMap,
 				Optional: true,
 				ForceNew: false,
+				Computed: true,
 			},
 			"snapshot_id": &schema.Schema{
 				Type:     schema.TypeString,

--- a/builtin/providers/openstack/resource_openstack_compute_instance_v2.go
+++ b/builtin/providers/openstack/resource_openstack_compute_instance_v2.go
@@ -209,7 +209,6 @@ func resourceComputeInstanceV2() *schema.Resource {
 					Schema: map[string]*schema.Schema{
 						"id": &schema.Schema{
 							Type:     schema.TypeString,
-							Optional: true,
 							Computed: true,
 						},
 						"volume_id": &schema.Schema{
@@ -955,7 +954,6 @@ func resourceComputeVolumeAttachmentHash(v interface{}) int {
 	var buf bytes.Buffer
 	m := v.(map[string]interface{})
 	buf.WriteString(fmt.Sprintf("%s-", m["volume_id"].(string)))
-	buf.WriteString(fmt.Sprintf("%s-", m["device"].(string)))
 	return hashcode.String(buf.String())
 }
 

--- a/builtin/providers/openstack/resource_openstack_compute_instance_v2_test.go
+++ b/builtin/providers/openstack/resource_openstack_compute_instance_v2_test.go
@@ -46,7 +46,7 @@ func TestAccComputeV2Instance_basic(t *testing.T) {
 	})
 }
 
-func _TestAccComputeV2Instance_volumeAttach(t *testing.T) {
+func TestAccComputeV2Instance_volumeAttach(t *testing.T) {
 	var instance servers.Server
 	var volume volumes.Volume
 
@@ -229,6 +229,7 @@ var testAccComputeV2Instance_volumeAttach = fmt.Sprintf(`
   resource "openstack_compute_instance_v2" "foo" {
     region = "%s"
     name = "terraform-test"
+    security_groups = ["default"]
     volume {
       volume_id = "${openstack_blockstorage_volume_v1.myvol.id}"
     }

